src/LocateDependencies/LocateDependenciesViaComposer.php

```diff
@@ -34,13 +37,11 @@ public function __construct(
 
     public function __invoke(string $installationPath) : SourceLocator
     {
-        Assert::that($installationPath)->directory();
         Assert::that($installationPath . '/composer.json')->file();
 
         $this->runInDirectory(function () use ($installationPath) : void {
             $installer = ($this->makeComposerInstaller)($installationPath);
 
-            Assert::that($installer)->isInstanceOf(Installer::class);
             assert($installer instanceof Installer);
```

Comment:
We can't really assume that `assert` will work on all platforms since it is configurable out of our control, so the `beberlei/assert` should remain, since the `makeComposerInstaller` property is just `callable` and has no defined type as the return value

Reply:
It is a `callable () : Installer` - that wouldn't type-check anymore, hence why I removed the assertion. Can re-introduce it, but can't write a test that triggers this assertion.

Reply:
Surely doing `new LocateDependenenciesViaComposer(function () {}, $astLocator)` is enough to trigger this scenario...?

Reply:
That doesn't type-check anymore, thanks to Psalm

Reply:
Doesn't mean it can't be done at runtime, by any consumer of the library....

Reply:
I'd rather say that consumers should respect the contract here: we've been doing this sort of assumption for ages before PHP 7, so I think we should keep doing it, especially since we now can provide explicit tooling for this

Reply:
(and yes, generic types in the engine would be better)

Reply:
We put strict types and assertions every where to avoid consumers mis-using the library and basically making the API foolproof. By removing stuff like this, you are trusting that all consumers understand what `callable () : Installer` actually means (I for example, haven't seen the syntax before, it's new and unclear, so I just assume "oh this needs a callable"). Whilst the annotations are good for static analysis, it's never enforced at runtime, so we need assertions still in place, and NOT using `assert` (because it is effectively useless due to it relying on consumer configuration).

Reply:
> We put strict types and assertions every where to avoid consumers mis-using the library and basically making the API foolproof. 

Consumers are still very much free to call our API without `declare(strict_types=1);` in their code, so the fool-proofing is mostly in our internal method calls.

> By removing stuff like this, you are trusting that all consumers understand what `callable () : Installer` actually means (I for example, haven't seen the syntax before, it's new and unclear, so I just assume "oh this needs a callable").

Absolutely clear that most people will not have seen this before, bit I think we (@Roave) are trend-setters, and we should push the ecosystem forward on this stuff. I did start pushing this into `doctrine/collections` too, hoping that the PHP ecosystem picks it up 👍

> Whilst the annotations are good for static analysis, it's never enforced at runtime

This is correct and intentional

> so we need assertions still in place

This is where we diverge: it is impossible to have a program that type-checks and triggers type assertions. In order to do that, I would have to write tests that intentionally do not respect the type declarations either.

Besides this technical limitation (which I believe to be correct, not an accident, but rather a happy combo of tooling discovering controversial definitions), I think that type checks are to be performed before runtime. Yes, the PHP engine is "special", because type checks are also applied at runtime, but IMO that shouldn't matter, if a program can be proven to be wrong before a single line of code is executed.



> and NOT using assert (because it is effectively useless due to it relying on consumer configuration).

Fully agree on this point: `assert()` is just bad, and as discussed before, I'll put effort into adding plugins for Psalm here.